proposals/007-azure-kms.md

```diff
@@ -0,0 +1,278 @@
+2. Support RSA and HSM-RSA key types, wrapping using `RSA-OAEP-256`.
+3. Support HSM-AES key type and AES-GCM wrapping
+4. Support only client credentials oauth flow with Entra using clientId + clientSecret. This supports workloads running anywhere. We could add support for other client credentials (certificates, federated certs) and Managed Identities later. Share a single auth token per Filter Definition until near expiry.
+5. Support TLS customization of the authentication client and key vault client.
+6. If the Key Vault is Managed HSM, then we will use the API to generate random bytes. Else, DEK bytes will be generated proxy-side with a SecureRandom.
+7. The Azure SDK pulls in netty/jackson/project-reactor, lets try implementing the APIs ourselves as we have for AWS
+8. [Edek](#edek-serialization-scheme) stores the keyName, keyVersion, edek. We attempt to minimise keyVersion size by optimistically decoding it from hex string, else store the string.
+9. User will supply tenantId for authentication, rather than implementing a more complicated workflow to obtain it using an HTTP request to KeyVault
+10. Endpoints for Entra and Key Vault will be configurable to support national clouds.
+11. It will be a documentation/user responsibility to ensure that their topic names are valid Azure key names if they
+    wish to use the `$(topicName)` KEK selector.
+
+### Key Types Support Matrix
+
+For completeness, this covers which key types are supported by this KMS implementation and which Key Vault products they
+are currently compatible with.
+
+| Azure key type | Supported | Wrapping Algorithm | Key Vault Standard | Key Vault Premium SKU | Managed HSM |
+|----------------|-----------|--------------------|--------------------|-----------------------|-------------|
+| RSA            | yes       | RSA-OAEP-256       | yes                | yes                   | yes         |
+| RSA-HSM        | yes       | RSA-OAEP-256       | no                 | yes                   | yes         |
+| oct            | yes       | AES-GCM            | no                 | no                    | yes         |
+| oct-HSM        | yes       | AES-GCM            | no                 | no                    | yes         |
+| EC             | no        | -                  | -                  | -                     | -           |
+| EC-HSM         | no        | -                  | -                  | -                     | -           |
+
+### Configuration
+
+```
+kms: AzureKeyVault
+kmsConfig:
+  keyVaultBaseUri: https://kv-kfesiehfoieaf.vault.azure.net
+  tls:
+    ... client trust for key vault
+  entraIdentity:
+     oauthEndpointUrl: https://login.microsoftonline.com // optional defaults to this value
+     clientId:
+       passwordFile: /path/to/id
+     clientSecret:
+       passwordFile: /path/to/id
+     tenantId: "abds-1232dsaa"
+     scope: https://vault.azure.net/.default // optional, could infer from vaultBaseUri
```

Comment:
Let's require it for now. That way people will be able to do funky things if they have use cases we haven't considered. Because:
* As with the tenant id, making it optional in the future will be a backwards compatible change, 
* No one can ever be blocked by their use case not being supported, so the improvement to make it smarter is work that readily deferrable. Whereas a work item to make it configurable might not be.